### PR TITLE
Add kubeconfig_content configuration parameter to the provider

### PR DESCRIPTION
**Pull request by melan**
Problem this change solves: 
If kubeconfig is stored somewhere (configuration servers, AWS SSM, etc.) but not on the file system then `terraform plan` will fail because the provider needs the config to check the state but the file can be created only when `plan` is done.

I hit this problem on AWS: I manage environments from a container in a build pipeline and kubeconfig  is generated by another module (https://github.com/terraform-aws-modules/terraform-aws-eks). This fix allows to pass variable with kubeconfig directly to the provider and complete both `plan` and `apply` phases successfully.

### CODE REVIEW

File: main.go [hunk 1]
```diff
@@ -69,26 +77,62 @@ func run(cmd *exec.Cmd) error {
...
-func kubectl(m interface{}, args ...string) *exec.Cmd {
+func kubectl(m interface{}, args ...string) (cmd *exec.Cmd, cleanupFunc func(), err error) {
 	kubeconfig := m.(*config).kubeconfig
+	kubeconfigContent := m.(*config).kubeconfigContent
+	if kubeconfigContent != "" {
+		content := []byte(kubeconfigContent)
+		var tmpfile *os.File
+		tmpfile, err = ioutil.TempFile("", "kubeconfig_")
+		if err != nil {
+			return
```
Comment (ericchiang) [inline main.go]: nit: please avoid the anonymous returns

```
tmpfile, err := ioutil.TempFile("", "kubeconfig_")
if err != nil {
    return nil, nil, fmt.Errorf("creating temp file for kubeconfig content: %v", err)
}
```

File: main.go [hunk 2]
```diff
@@ -69,26 +77,62 @@ func run(cmd *exec.Cmd) error {
...
+		}
+
+		cleanupFunc = func() { os.Remove(tmpfile.Name()) }
+
+		if _, err = tmpfile.Write(content); err != nil {
```
Comment (ericchiang) [inline main.go]: nit: io.WriteString

```
if _, err := io.WriteString(tempfile, content); err != nil {
    return fmt.Errorf("writing kubeconfig to file: %v", err)
}
```

**Comment by whyman** [comment]
:+1:  We need this please!

File: main.go [hunk 3]
```diff
@@ -69,26 +77,62 @@ func run(cmd *exec.Cmd) error {
 	return nil
 }
 
-func kubectl(m interface{}, args ...string) *exec.Cmd {
+func kubectl(m interface{}, args ...string) (cmd *exec.Cmd, cleanupFunc func(), err error) {
```
Comment (ericchiang) [inline main.go]: How about this. Just pass the kubectl function a kubeconfig path?

```go
func kubectl(m interface{}, kubeconfig string, args ...string) (*exec.Cmd, error) 
```

Then create a helper

```go
func kubeconfigPath(m interface{}) (string, func(), error)
```

Then blocks like this

```go
	cmd := kubectl(m, "apply", "-f", "-")
	cmd.Stdin = strings.NewReader(d.Get("content").(string))
	if err := run(cmd); err != nil {
		return err
	}

	stdout := &bytes.Buffer{}
	cmd = kubectl(m, "get", "-f", "-", "-o", "json")
	cmd.Stdin = strings.NewReader(d.Get("content").(string))
	cmd.Stdout = stdout
	if err := run(cmd); err != nil {
		return err
	}
```

Become

```go
	kubeconfig, cleanup, err := kubeconfigPath(m)
	if err != nil {
		return fmt.Errorf("determining kubeconfig: %v", err)
	}
	defer cleanup()

	cmd := kubectl(m, kubeconfig, "apply", "-f", "-")
	cmd.Stdin = strings.NewReader(d.Get("content").(string))
	if err := run(cmd); err != nil {
		return err
	}

	stdout := &bytes.Buffer{}
	cmd = kubectl(m, kubeconfig, "get", "-f", "-", "-o", "json")
	cmd.Stdin = strings.NewReader(d.Get("content").(string))
	cmd.Stdout = stdout
	if err := run(cmd); err != nil {
		return err
	}
```

**Review by ericchiang — COMMENTED** [review]
Thanks for the change! Left some comments.

**Comment by melan** [comment]
@ericchiang I updated the code, thanks for your feedback

File: README.md [hunk 4]
```diff
@@ -18,14 +18,27 @@ providers {
 }
 ```
 
-The provider takes optional configuration to specify a `kubeconfig` file:
+The provider takes couple optional configuration parameters:
```
Comment (ericchiang) [inline README.md]: spelling mistake here, maybe

"takes the following optional configuration parameters"

File: main.go [hunk 5]
```diff
@@ -69,23 +78,61 @@ func run(cmd *exec.Cmd) error {
 	return nil
 }
 
-func kubectl(m interface{}, args ...string) *exec.Cmd {
+func kubeconfigPath(m interface{}) (string, func(), error) {
 	kubeconfig := m.(*config).kubeconfig
+	kubeconfigContent := m.(*config).kubeconfigContent
+	var cleanupFunc func() = nil
```
Comment (ericchiang) [inline main.go]: cleanupFunc := func() {}

File: main.go [hunk 6]
```diff
@@ -114,7 +161,15 @@ func resourceManifestCreate(d *schema.ResourceData, m interface{}) error {
 }
 
 func resourceManifestUpdate(d *schema.ResourceData, m interface{}) error {
-	cmd := kubectl(m, "apply", "-f", "-")
+	kubeconfig, cleanup, err := kubeconfigPath(m)
+	if err != nil {
+		return fmt.Errorf("determining kubeconfig: %v", err)
+	}
+	if cleanup != nil {
```
Comment (ericchiang) [inline main.go]: cleanup should never be nil

File: main.go [hunk 7]
```diff
@@ -144,7 +199,15 @@ func resourceManifestDelete(d *schema.ResourceData, m interface{}) error {
 	if namespace != "" {
 		args = append(args, "-n", namespace)
 	}
-	return run(kubectl(m, args...))
+	kubeconfig, cleanup, err := kubeconfigPath(m)
+	if err != nil {
+		return fmt.Errorf("determining kubeconfig: %v", err)
+	}
+	if cleanup != nil {
```
Comment (ericchiang) [inline main.go]: ditto

File: main.go [hunk 8]
```diff
@@ -159,7 +222,14 @@ func resourceManifestRead(d *schema.ResourceData, m interface{}) error {
 	}
 
 	stdout := &bytes.Buffer{}
-	cmd := kubectl(m, args...)
+	kubeconfig, cleanup, err := kubeconfigPath(m)
+	if err != nil {
+		return fmt.Errorf("determining kubeconfig: %v", err)
+	}
+	if cleanup != nil {
```
Comment (ericchiang) [inline main.go]: ditto

File: main.go [hunk 9]
```diff
@@ -69,23 +78,61 @@ func run(cmd *exec.Cmd) error {
...
+		cleanupFunc = func() { os.Remove(tmpfile.Name()) }
+
+		if _, err = io.WriteString(tmpfile, kubeconfigContent); err != nil {
+			return "", cleanupFunc, fmt.Errorf("writing kubeconfig to file: %v", err)
+		}
+		if err = tmpfile.Close(); err != nil {
+			return "", cleanupFunc, fmt.Errorf("completion of write to kubeconfig file: %v", err)
+		}
+
+		kubeconfig = tmpfile.Name()
```
Comment (ericchiang) [inline main.go]: it should be an error to specify both kubeconfig and kubeconfig_content

File: README.md [hunk 10]
```diff
@@ -18,14 +18,27 @@ providers {
...
 
+* If you content of the kubeconfig is available in a variable, you can configure the provider as:
+
+```hcl
+provider "k8s" {
+  kubeconfig_content = "${var.kubeconfig}"
+}
+```
+
+Configuration from the variable will be recorded into a temporary file and the file will be removed as
```
Comment (ericchiang) [inline README.md]: Make this a warning.

"WARNING: ..."

Writing to the global tempdir would be a pretty terrifying thing to do on a shared system.

File: main.go [hunk 11]
```diff
@@ -114,7 +162,13 @@ func resourceManifestCreate(d *schema.ResourceData, m interface{}) error {
 }
 
 func resourceManifestUpdate(d *schema.ResourceData, m interface{}) error {
-	cmd := kubectl(m, "apply", "-f", "-")
+	kubeconfig, cleanup, err := kubeconfigPath(m)
+	defer cleanup()
```
Comment (ericchiang) [inline main.go]: When `err != nil`, it's normally not safe to use the other returned arguments. For example, the following http.Client code will panic if an error's returned

```
resp, err := http.Get("http://example.com/")
defer resp.Body.Close()
if err != nil {
	// handle error
}
```

Please switch this to

```
kubeconfig, cleanup, err := kubeconfigPath(m)
if err != nil {
    return err
}
defer cleanup()
```

If you're concerned about cleanup always being called then kubeconfigPath should call it itself if it's going to return an error.

**Review by ericchiang — COMMENTED** [review]
last couple nits

**Comment by ericchiang** [comment]
Also, please squash your commits! :)

**Comment by melan** [comment]
@ericchiang can you take a look and merge it if everything is fine now?

**Pull request merged**